primitives/src/merkle_tree/field_based_mht/smt/big_lazy_merkle_tree.rs

```diff
@@ -0,0 +1,988 @@
+use crate::merkle_tree::field_based_mht::smt::{SmtPoseidonParameters, Coord, OperationLeaf, BigMerkleTreeState};
+use crate::{PoseidonParameters, BatchFieldBasedHash, merkle_tree};
+use crate::crh::poseidon::batched_crh::PoseidonBatchHash;
+use crate::merkle_tree::field_based_mht::smt::ActionLeaf::Remove;
+use crate::merkle_tree::field_based_mht::smt::parameters::{MNT4753SmtPoseidonParameters, MNT6753SmtPoseidonParameters};
+use crate::crh::poseidon::parameters::{MNT4753PoseidonParameters, MNT6753PoseidonParameters};
+
+use rocksdb::{DB, Options};
+
+use std::collections::HashSet;
+use std::marker::PhantomData;
+
+use std::fs;
+
+use algebra::{PrimeField, MulShort};
+use algebra::{ToBytes, to_bytes, FromBytes};
+use algebra::fields::mnt6753::Fr as MNT6753Fr;
+use algebra::fields::mnt4753::Fr as MNT4753Fr;
+use crate::merkle_tree::field_based_mht::smt::big_merkle_tree::BigMerkleTree;
+use crate::merkle_tree::field_based_mht::smt::error::Error;
+use std::path::Path;
+
+#[derive(Debug)]
+pub struct LazyBigMerkleTree<F: PrimeField + MulShort, T: SmtPoseidonParameters<Fr=F>, P: PoseidonParameters<Fr=F>> {
+    // if unset, all DBs and tree internal state will be deleted when an instance of this struct
+    // gets dropped
+    persistent: bool,
+    // path to state required to restore the tree
+    state_path: Option<String>,
+    // tree in-memory state
+    state: BigMerkleTreeState<F, T>,
+    // the height of the Merkle tree
+    height: usize,
+    // path to the db
+    path_db: String,
+    // stores the leaves
+    database: DB,
+    // path to the cache
+    path_cache: String,
+    // stores the cached nodes
+    db_cache: DB,
+
+    _field: PhantomData<F>,
+    _parameters: PhantomData<T>,
+    _poseidon_parameters: PhantomData<P>,
+}
+
+impl<F: PrimeField + MulShort, T: SmtPoseidonParameters<Fr=F>, P: PoseidonParameters<Fr=F>> Drop for LazyBigMerkleTree<F, T, P> {
+    fn drop(&mut self) {
+        self.close()
+    }
+}
+
+impl<F: PrimeField + MulShort, T: SmtPoseidonParameters<Fr=F>, P: PoseidonParameters<Fr=F>> LazyBigMerkleTree<F, T, P> {
+    // Creates a new tree of specified `width`.
+    // If `persistent` is specified, then DBs will be kept on disk and the tree state will be saved
+    // so that the tree can be restored any moment later. Otherwise, no state will be saved on file
+    // and the DBs will be deleted.
+    pub fn new_unitialized(width: usize, persistent: bool, state_path: Option<String>, path_db: String, path_cache: String) -> Result<Self, Error> {
+
+        // If the tree must be persistent, that a path to which save the tree state must be
+        // specified.
+        if !persistent { assert!(state_path.is_none()) } else { assert!(state_path.is_some()) }
+
+        //TODO: This is the "depth" more than the "height". Which should we keep ?
```

Comment:
This is the depth more than the height. Probably we should keep the notion of height, i.e. the number of levels from leaves to the root, and the number of leaves will be 2^(HEIGHT-1)